lib/HLSL/DxilValidation.cpp

```diff
@@ -1131,6 +1131,111 @@ static int GetCBufSize(Value *cbHandle, ValidationContext &ValCtx) {
   return RP.CBufferSizeInBytes;
 }
 
+void ValidateHandleArgsForInstruction(CallInst* CI, DXIL::OpCode opcode, ValidationContext &ValCtx) {
+  // make sure none of the handle arguments are undef / zero-initializer,
+  // The only exception is that a zero-initializer is allowed
+  // as an argument to Output Complete (CAZ == zero-initializer)
+  // Also, do not accept any resource handles with invalid dxil resource properties
+  for (Value *op : CI->operands()) {
+    const Type *pHandleTy = ValCtx.HandleTy; // This is a resource handle
+    const Type *pNodeHandleTy = ValCtx.DxilMod.GetOP()->GetNodeHandleType();
+    const Type *pNodeRecordHandleTy = ValCtx.DxilMod.GetOP()->GetNodeRecordHandleType();
+
+    const Type *argTy = op->getType();
+    if (argTy == pNodeHandleTy || argTy == pNodeRecordHandleTy || argTy == pHandleTy) {
+      bool bIsUninitialized = isa<UndefValue>(op) || isa<ConstantAggregateZero>(op);
+      if (argTy == pHandleTy) {
+        hlsl::DxilResourceProperties pDRP = ValCtx.GetResourceFromVal(op);
```

Comment:
- This should use `GetResourceFromHandle` instead.
- The name `pDRP` implies a pointer, but this isn't a pointer.
- There's no need to emit another error when `pDRP` is invalid, since `GetResourceFromHandle` will emit `InstrHandleNotFromCreateHandle`
- We should update the error emitted from `InstrHandleNotFromCreateHandle`, which is currently: `"Resource handle should returned by createHandle."`.  Grammar issue aside, the error is wrong on SM 6.6+ where the handle should come from annotateHandle instead of createHandle (and createHandleForLib for lib targets < 6.6).

Reply:
Thanks, good comments. I only have one problem:
There is a difference when switching from GetResourceFromHandle and GetResourceFromVal.
The ResourceClass is tested for validity in the GetResourceFromHandle case, while the ResourceKind is tested for validity in the GetResourceFromVal case. I've switched over to GetResourceFromHandle, but left the .isValid check in since it checks for a different validity than what's vaidated within the function itself.
Unless there's some assumption we can make that ResourceKind will be invalid if and only if ResourceClass is, I think the check should be left.

Reply:
`GetResourceFromVal()` doesn't check for validity at all, which is why you needed the .IsValid() check.
`GetResourceFromHandle()` calls `DxilResourceProperties::getResourceClass()`, which checks the `ResourceKind`, and will return `ResourceClass::Invalid` when `ResourceKind` is `Invalid` (same condition as for `IsValid()`).  See my new comment which points this out and more.

Earlier, I missed following up on this because you marked the issue as resolved, and it disappeared from the code due to changes.  If there's a response like this and it may not have fully resolved all concerns, please don't resolve the comment, so we don't lose track of it.